File: 1250-longest-common-subsequence/longest-common-subsequence.ex
defmodule Solution do
  @spec longest_common_subsequence(text1 :: String.t, text2 :: String.t) :: integer
  def longest_common_subsequence(text1, text2) do
    text1 = String.to_charlist(text1)
    text2 = String.to_charlist(text2)

    dp =
      text1
      |> Enum.with_index()
      |> Enum.reduce(%{}, fn {c1, c1i}, dp ->
        text2
        |> Enum.with_index()
        |> Enum.reduce(dp, fn {c2, c2i}, dp ->
          if c1 == c2 do
            Map.put(dp, {c1i, c2i}, Map.get(dp, {c1i - 1, c2i - 1}, 0) + 1)
          else
            Map.put(
              dp,
              {c1i, c2i},
              max(Map.get(dp, {c1i - 1, c2i}, 0), Map.get(dp, {c1i, c2i - 1}, 0))
            )
          end
        end)
      end)

    Map.get(dp, {length(text1) - 1, length(text2) - 1})
  end
end

File: 1250-longest-common-subsequence/longest-common-subsequence.swift
class Solution {
    func longestCommonSubsequence(_ text1: String, _ text2: String) -> Int {
        var m = text1.count, n = text2.count
        var dp = Array(repeating: Array(repeating:0, count: n+1), count: m+1)
        var arr1 = Array(text1), arr2 = Array(text2)

        for i in 1...m {
            for j in 1...n {
                if arr1[i-1] == arr2[j-1] {
                    dp[i][j] = dp[i-1][j-1] + 1
                } else {
                    dp[i][j] = max(dp[i-1][j], dp[i][j-1])
                }
            }
        }

        return dp[m][n]
    }
}